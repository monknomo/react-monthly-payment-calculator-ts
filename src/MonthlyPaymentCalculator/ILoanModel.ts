export default interface ILoanModel {
  vehiclePrice: number;
  tax: number;
  fees: number;
  downPayment?: number;
  apr?: number;
  term?:number;
}
