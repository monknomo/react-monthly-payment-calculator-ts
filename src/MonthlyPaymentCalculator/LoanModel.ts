import ILoanModel from './ILoanModel';

export default class LoanModel {

  public static precisionRound(unroundedNumber: number, precision:number ): number {
    const factor = Math.pow(10, precision);
    return Math.round(unroundedNumber * factor) / factor;
  }

  constructor(public loanModel: ILoanModel){
    const errors: string[] = [];
    this.validateLoanModel(loanModel, errors);
    if(errors.length > 0){
      throw new Error(this.errorsToString(errors));
    }
    this.loanModel = loanModel;

  }

  // monthly payment = (monthlyInterestRate * principal) / (1 - (1 + monthlyInterestRate)^-term)

  public calculateMonthlyPayment(): number {
    const monthlyInterestRate = this.calculateMonthlyInterest();
    const principal = this.calculatePrincipal();
    if(this.loanModel.term === null || this.loanModel.term === undefined){
      throw new Error("Loan term is required");
    }
    if(monthlyInterestRate ===0){
      return LoanModel.precisionRound(principal / this.loanModel.term, 2);
    }
    const payment =  (monthlyInterestRate*principal) / (1-Math.pow((1+monthlyInterestRate), -1*this.loanModel.term));
    return LoanModel.precisionRound(payment, 2);
  }

  public calculatePrincipal(): number {
    if(this.loanModel.downPayment){
      return LoanModel.precisionRound((this.loanModel.vehiclePrice * (1 + this.loanModel.tax / 100) +this.loanModel.fees)-this.loanModel.downPayment, 2);
    }
    return LoanModel.precisionRound(this.loanModel.vehiclePrice * (1 + this.loanModel.tax / 100) +this.loanModel.fees, 2);

  }

  public calculateMonthlyInterest(): number{
    debugger;
    if(this.loanModel.apr === null || this.loanModel.apr === undefined || isNaN(this.loanModel.apr) ){
      throw new Error("APR is required");
    }
    return this.loanModel.apr / 100 / 12;
  }

  public validateLoanModel(loanModel: ILoanModel, errors: string[]): void{
    this.validateVehiclePrice(loanModel, errors);
    this.validateTax(loanModel, errors);
    this.validateFees(loanModel, errors);
    this.validateDownPayment(loanModel, errors);
    this.validateApr(loanModel, errors);
    this.validateTerm(loanModel, errors);
  }

  public validateTerm(loanModel: ILoanModel, errors: string[]): void{
    this.termMustBePositive(loanModel, errors);
    this.termLessThanTenYears(loanModel, errors);
  }

  public validateApr(loanModel: ILoanModel, errors: string[]): void{
    this.aprMustBePositive(loanModel, errors);
    this.aprMustBeLessThanOneHundred(loanModel, errors);
  }

  public validateDownPayment(loanModel: ILoanModel, errors: string[]): void{
    this.downPaymentMustBePositive(loanModel, errors);
    this.downPaymentLessThanPrice(loanModel, errors);
  }

  public validateFees(loanModel: ILoanModel, errors: string[]): void{
    this.feesMustBePositive(loanModel, errors);
    this.feesLessThanMax(loanModel, errors);
  }

  public validateTax(loanModel: ILoanModel, errors: string[]): void{
    this.taxMustBeLessThanMax(loanModel, errors);
    this.taxMustBePositive(loanModel, errors);
  }

  public validateVehiclePrice(loanModel: ILoanModel, errors: string[]): void{
    this.vehiclePriceMustBePositive(loanModel, errors);
    this.vehiclePriceMustBeLessThanMax(loanModel, errors);
  }

  public vehiclePriceMustBePositive(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.vehiclePrice< 0){
      errors.push("Positive dollar amount expected for vehicle price\n");
    }
  }

  public vehiclePriceMustBeLessThanMax(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.vehiclePrice> 500000.00){
      errors.push("Vehicle price too high\n");
    }
  }

  public taxMustBeLessThanMax(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.tax>100.00){
      errors.push("Tax too high\n");
    }
  }

  public taxMustBePositive(loanModel:ILoanModel, errors: string[]): void{
    if(loanModel.tax<0){
      errors.push("Tax must be positive percentage\n");
    }
  }

  public feesMustBePositive(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.fees<0){
      errors.push("Fees must be positive dollar amount\n");
    }
  }

  public feesLessThanMax(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.fees>10000.00){
      errors.push("Fees too high\n");
    }
  }

  public downPaymentMustBePositive(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.downPayment && loanModel.downPayment < 0){
      errors.push("Down payment must be a positive dollar amount\n");
    }
  }

  public downPaymentLessThanPrice(loanModel: ILoanModel, errors: string[]): void{
    debugger;
    if(loanModel.downPayment && loanModel.downPayment > loanModel.vehiclePrice){
      errors.push("Down payment must be less than vehicle price\n");
    }
  }

  public aprMustBePositive(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.apr && loanModel.apr < 0){
      errors.push("APR must be a positive percentage\n");
    }
  }

  public aprMustBeLessThanOneHundred(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.apr && loanModel.apr>100){
      errors.push("APR too high, seek better loan\n");
    }
  }

  public termMustBePositive(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.term && loanModel.term<0){
      errors.push("Loan term must be a positive number of months\n");
    }
  }

  public termLessThanTenYears(loanModel: ILoanModel, errors: string[]): void{
    if(loanModel.term && loanModel.term>120){
      errors.push("Loan term too long\n");
    }
  }

  private errorsToString(errors: string[]): string{
    const errorMessage: string = errors.join(',');
    return errorMessage.substring(0, errorMessage.length - 1);
  }
}
