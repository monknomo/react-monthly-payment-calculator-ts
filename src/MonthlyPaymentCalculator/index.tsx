import * as React from 'react';
import MaskedInput from 'react-text-mask';
import {createNumberMask} from 'text-mask-addons';
import ILoanModel from './ILoanModel';
import LoanModel from './LoanModel';

interface IMonthlyPaymentCalculatorProps {
  vehiclePrice: number,
  tax: number,
  fees:number
};

interface IMonthlyPaymentCalculatorState {
  apr: string,
  downPayment: string,
  term: string
};

export default class MonthlyPaymentCalculator extends React.Component<IMonthlyPaymentCalculatorProps,IMonthlyPaymentCalculatorState>{

  constructor(props: IMonthlyPaymentCalculatorProps){
    super(props);
    this.state = {apr:"0450", downPayment:"0",term:"60"};
    this.handleChange = this.handleChange.bind(this);
  }

  public handleChange(event: any): void {
    const field = event.target.name;
    const modifiedState = {};
    modifiedState[field] = event.target.value;
    this.setState(modifiedState);
  }

  public render(){
    const aprPercent = Number(this.state.apr.substring(0,2).replace(/\_/g,""));
    const aprDecimal =  Number(this.state.apr.substring(2,4).replace(/\_/g,""))/10;
    let numericApr = aprPercent;
    if(!isNaN(aprDecimal)){
      numericApr += aprDecimal;
    }
    const loanModel: ILoanModel = {
      apr: numericApr,
      downPayment: Number(this.state.downPayment.replace(/\$|\_|,/g, "")),
      fees: this.props.fees,
      tax: this.props.tax,
      term: Number(this.state.term),
      vehiclePrice: this.props.vehiclePrice
    }
    let errors = "";
    let monthlyPayment: number = 0;
    let principal: number = 0;
    try{
      const paymentMath: LoanModel = new LoanModel(loanModel);
      monthlyPayment = paymentMath.calculateMonthlyPayment();
      principal = paymentMath.calculatePrincipal();
    }catch(e){
      errors = e.message;
    }
    return (
      <form>

        <div>
          <h1>Car Loan Monthly Payment Calculator</h1>
        </div>
        <div>
          {errors}
        </div>
        <div>
          Vehicle Price: ${this.props.vehiclePrice}
        </div>
        <div>
          <label htmlFor="downPayment">Down Payment: </label>
          <MaskedInput name="downPayment" mask={createNumberMask({})}
            id="monthlyPaymentCalculator-downPayment" required={true}
            value={this.state.downPayment} onChange={this.handleChange}/>

        </div>
        <div>
          <label htmlFor="apr">APR: </label>
          <MaskedInput name="apr" mask={[/\d/,/\d/,'.',/\d/,/\d/,'%']} required={true}
            value={this.state.apr} onChange={this.handleChange}/>
        </div>
        <div>
          <label htmlFor="term">Term: </label>
          <select name="term" id="monthlyPaymentCalculator-term"
            value={this.state.term} onChange={this.handleChange}>
            <option value="24">24 Month Term</option>
            <option value="36">36 Month Term</option>
            <option value="48">48 Month Term</option>
            <option value="60">60 Month Term</option>
          </select>
        </div>
        <div>
          <h2>Estimated Payment</h2>
          <div>Tax Rate: {this.props.tax}%</div>
          <div>Total Tax: ${this.props.vehiclePrice * this.props.tax / 100}</div>
          <div>Fees: ${this.props.fees}</div>
          <div>Principal: ${principal}</div>
          <hr/>
          <div>Monthly Payment: ${monthlyPayment}</div>
        </div>
    </form>
    )
  }

}
