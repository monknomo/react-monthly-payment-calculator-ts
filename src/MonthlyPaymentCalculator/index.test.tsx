import * as React from 'react';
import * as ReactDOM from 'react-dom';
import MonthlyPaymentCalculator from './index';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MonthlyPaymentCalculator  vehiclePrice={25000} tax={10.4} fees={242}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});
