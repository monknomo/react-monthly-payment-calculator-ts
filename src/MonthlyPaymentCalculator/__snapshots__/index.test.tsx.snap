// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<form>
  <div>
    <h1>
      Car Loan Monthly Payment Calculator
    </h1>
  </div>
  <div>
    
  </div>
  <div>
    Vehicle Price: $
    25000
  </div>
  <div>
    <label
      htmlFor="downPayment"
    >
      Down Payment: 
    </label>
    <input
      id="monthlyPaymentCalculator-downPayment"
      mask=""
      name="downPayment"
      onChange={[Function]}
      required={true}
      value="0"
    />
  </div>
  <div>
    <label
      htmlFor="apr"
    >
      APR: 
    </label>
    <input
      mask={
        Array [
          /\\\\d/,
          /\\\\d/,
          ".",
          /\\\\d/,
          /\\\\d/,
          "%",
        ]
      }
      name="apr"
      onChange={[Function]}
      required={true}
      value="0450"
    />
  </div>
  <div>
    <label
      htmlFor="term"
    >
      Term: 
    </label>
    <select
      id="monthlyPaymentCalculator-term"
      name="term"
      onChange={[Function]}
      value="60"
    >
      <option
        value="24"
      >
        24 Month Term
      </option>
      <option
        value="36"
      >
        36 Month Term
      </option>
      <option
        value="48"
      >
        48 Month Term
      </option>
      <option
        value="60"
      >
        60 Month Term
      </option>
    </select>
  </div>
  <div>
    <h2>
      Estimated Payment
    </h2>
    <div>
      Tax Rate: 
      10.4
      %
    </div>
    <div>
      Total Tax: $
      2600
    </div>
    <div>
      Fees: $
      242
    </div>
    <div>
      Principal: $
      27842
    </div>
    <hr />
    <div>
      Monthly Payment: $
      577.95
    </div>
  </div>
</form>
`;
